feat(curriculum): date conversion lab (#56332)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-date-conversion/66f686b8ebdb982fa8e14330.md

2. You should create a variable named `currentDateFormat` and assign it the string `Current Date and Time: [current date]`. Replace `[current date]` with the result of the `currentDate` variable.
3. You should log the value of `currentDateFormat` to the console.
4. You should create a function named `formatDateMMDDYYYY` that takes the date object as a parameter. You can name this parameter whatever you like.
5. Your `formatDateMMDDYYYY` function should convert the current date to `month/day/year` format and return the string `Formatted Date (MM/DD/YYYY): [month]/[day]/[year]`.
6. You should create a function named `formatDateLong` that takes the date object as a parameter.
7. Your `formatDateLong` function should convert the current date to `Month Day, Year` format and return the string `Formatted Date (Month Day, Year): [formatted date]`.

# --hints--

You should have a variable named `currentDate` that holds the current date and time using the `Date` object.

```js

// excluded timestamps
const expectedDate = new Date();
const expectedDateOnly = new Date(expectedDate.getFullYear(), expectedDate.getMonth(), expectedDate.getDate());

> const currentDateOnly = new Date(currentDate.getFullYear(), currentDate.getMonth(), currentDate.getDate());

Reviewer: why did you create these but did not use them?

> assert.equal(currentDate.toString(), expectedDate.toString());

Reviewer 2: ```suggestion
assert.equal(currentDateOnly.toString(), expectedDateOnly.toString());
```
I guess this was the idea.

Author: Yes, exactly this is what was required.

```

You should have a variable named `currentDateFormat` that holds the current date in the format `Current Date and Time: [current date]`.

```js
const expectedDate = new Date();
const expectedDateFormat = `Current Date and Time: ${expectedDate}`;

assert.equal(currentDateFormat, expectedDateFormat);
```

You should log the value of `currentDateFormat` to the console.

```js
assert.match(code, /console\.log\(\s*currentDateFormat\s*\)/);
```

You should have a function `formatDateMMDDYYYY` 

```js
assert.isFunction(formatDateMMDDYYYY);
```

> The function`formatDateMMDDYYYY` should take a date object as a parameter.
> 
> ```js
> assert.match(formatDateMMDDYYYY.toString(), /\s*formatDateMMDDYYYY\(\s*\w+\s*\)/);
> ```

Reviewer: this is not testing that the parameter is a date object. To test if it has one parameter you can check `formatDateMMDDYYY.length` (I learned recently that the lenght is equal to the number of parameters)

````suggestion
The function`formatDateMMDDYYYY` should take one parameter.

```js
assert.lengthOf(formatDateMMDDYYYY, 1);
```
````

When the date object holds `Fri Sep 27 2024 16:16:11 GMT+0500 (Pakistan Standard Time)`, the function `formatDateMMDDYYYY` should return `Formatted Date (MM/DD/YYYY): 9/27/2024`.

```js
let dateString = "Fri Sep 27 2024 16:21:18 GMT+0500 (Pakistan Standard Time)";
let dateObject = new Date(dateString);

assert.equal(formatDateMMDDYYYY(dateObject).toString(), "Formatted Date (MM/DD/YYYY): 9/27/2024");

// prevent hardcoding

dateString = "Fri Sep 28 2024 18:31:18 GMT+0500 (Pakistan Standard Time)";
dateObject = new Date(dateString);

assert.strictEqual(formatDateMMDDYYYY(dateObject).toString(), "Formatted Date (MM/DD/YYYY): 9/28/2024");
```

You should have a function `formatDateLong` 

```js
assert.isFunction(formatDateLong);
```

The function`formatDateLong` should take a date object as a parameter.

```js

> assert.match(formatDateLong.toString(), /\s*formatDateLong\(\s*\w+\s*\)/);

Reviewer: how is this testing that the parameter is a date object?

Reviewer: ```js
assert.lengthOf(formatDateLong, 1);
```

Author: It is not checking specifically for the date object, I have updated the hint text.

**Note:** For the tests to pass, make sure that you use `en-US` for the locale when formatting the dates.

# --hints--

You should have a variable named `currentDate` that holds the current date and time using the `Date` object.

```js
// excluded timestamps
const expectedDate = new Date();
const expectedDateOnly = new Date(expectedDate.getFullYear(), expectedDate.getMonth(), expectedDate.getDate());

const currentDateOnly = new Date(currentDate.getFullYear(), currentDate.getMonth(), currentDate.getDate());

assert.equal(currentDateOnly.toString(), expectedDateOnly.toString());
```

You should have a variable named `currentDateFormat` that holds the current date in the format `Current Date and Time: [current date]`.

```js
const expectedDate = new Date();
const expectedDateFormat = `Current Date and Time: ${expectedDate}`;

assert.equal(currentDateFormat, expectedDateFormat);
```

You should log the value of `currentDateFormat` to the console.

```js
assert.match(code, /console\.log\(\s*currentDateFormat\s*\)/);
```

You should have a function `formatDateMMDDYYYY` 

```js
assert.isFunction(formatDateMMDDYYYY);
```

> The function`formatDateMMDDYYYY` should take a date object as a parameter.

Reviewer 2: ```suggestion
The function`formatDateMMDDYYYY` should take a single parameter.
```
